### ISSUE TITLE
Script: Cannot Schedule Job using Script Resource

### ISSUE THREAD

**Issue by Chirishman** [PowerShell/PSDscResources#91]
I've been trying to use a DSC Script resource to toggle the `RebootNodeIfNeeded` flag in LCM based on datetime so that I can control the patching reboot window.

Since I don't seem to be able to make the execution of one configuration step dependent on another or replace the boolean true or false setting for `RebootNodeIfNeeded` with an evaluation, I've fallen back to trying to create a powershell scheduled job to define, generate, and apply metaconfig MOF.

Unfortunately I am somehow getting an elevation related error when the below configuration hits the Register-ScheduledJob step in the SetScript.

```
Configuration UpdateAndScheduleReboot{
 
    Param(
        [string]$NodeName='localhost',
        [datetime]$TargetRebootTime
    )
    
    Import-DscResource –ModuleName ’PSDesiredStateConfiguration’
    Import-DSCResource -ModuleName xPendingReboot
    Import-DSCResource -ModuleName xWindowsUpdate
    $TargetRebootTime = $AllNodes.TargetRebootTime
    Script RebootEnableJob {
        GetScript = {
            return @{Result = "HelloWorld"}
        }
        TestScript = {
            $ScheduledEnableJob = Get-ScheduledJob | ?{$_.Name -eq 'EnableReboot'}
            if ($ScheduledEnableJob){
                if ($true <#Check for time match#>){
                    Write-Verbose "Matching job found"
                    return $true
                } else {
                    Write-Verbose "Matching job found - Stale Trigger Time"
                    return $false
                }

            } else {
                Write-Verbose "No matching job found"
                Return $false
            }
        }
        SetScript = {
            $ScheduledEnableJob = Get-ScheduledJob | ?{$_.Name -eq 'EnableReboot'}
            if ($ScheduledEnableJob){
                Write-Verbose "Modifying Existing Job"
                $ScheduledEnableJob | Set-ScheduledJob -Trigger (New-JobTrigger -Once -At $Using:TargetRebootTime)
            } else {
                Write-Verbose "Creating New Scheduled Job with target trigger time $($Using:TargetRebootTime)"
                Write-Verbose "User: $([System.Security.Principal.WindowsIdentity]::GetCurrent().Name)"
                Write-Verbose "Current User: $env:UserName"
                $EnableJob = @{
                    Name = 'EnableReboot'
                    ScriptBlock = {
                        configuration RebootSetting
                        {
	                        Param(
		                        [string[]]$NodeName = 'localhost'
	                        )
                            LocalConfigurationManager {
                                RebootNodeIfNeeded = $true
                            }
                        }
                        $ConfigPath = "C:\temp\RebootSetting"
                        & ([scriptblock]::Create("RebootSetting -OutputPath $ConfigPath"))
                        Set-DscLocalConfigurationManager -Path $ConfigPath
        
                    }
                    Trigger = New-JobTrigger -Once -At $Using:TargetRebootTime
                    ScheduledJobOption = New-ScheduledJobOption -RunElevated -DoNotAllowDemandStart
                    ErrorAction = 'Stop'
                }
                Register-ScheduledJob @EnableJob
            }
        }
    }

    Script RebootDisableJob {
        GetScript = {
            return @{Result = "HelloWorld"}
        }
        TestScript = {
            $ScheduledDisableJob = Get-ScheduledJob | ?{$_.Name -eq 'DisableReboot'}
            if ($ScheduledDisableJob){
                if ($true <#Check for time match#>){
                    Write-Verbose "Matching job found"
                    return $true
                } else {
                    Write-Verbose "Matching job found - Stale Trigger Time"
                    return $false
                }

            } else {
                Write-Verbose "No matching job found"
                Return $false
            }
        }
        SetScript = {
            $ScheduledDisableJob = Get-ScheduledJob | ?{$_.Name -eq 'DisableReboot'}
            if ($ScheduledDisableJob){
                Write-Verbose "Modifying Existing Job"
                $ScheduledDisableJob | Set-ScheduledJob -Trigger (New-JobTrigger -Once -At ($Using:TargetRebootTime).AddHours(1.5))
            } else {
                Write-Verbose "Creating New Scheduled Job with target trigger time $Using:TargetRebootTime"
                $DisableJob = @{
                    Name = 'DisableReboot'
                    ScriptBlock = {
                        configuration RebootSetting
                        {
	                        Param(
		                        [string[]]$NodeName = 'localhost'
	                        )
                            LocalConfigurationManager {
                                RebootNodeIfNeeded = $False
                            }
                        }
                        $ConfigPath = "C:\temp\RebootSetting"
                        "RebootSetting -OutputPath $ConfigPath"
                        Set-DscLocalConfigurationManager -Path $ConfigPath
        
                    }
                    Trigger = New-JobTrigger -Once -At $Using:TargetRebootTime
                    ScheduledJobOption = New-ScheduledJobOption -RunElevated -DoNotAllowDemandStart
                }
                Register-ScheduledJob @DisableJob
            }
        }
    }

    xWindowsUpdateAgent MuSecurityImportant {
        IsSingleInstance = 'Yes'
        UpdateNow        = $True
        Source           = 'WindowsUpdate'
        Notifications    = 'ScheduledInstallation'
    }

    xPendingReboot CheckReboot {
        Name = "Check for a pending reboot and trigger if true"
    }
 
}

$ConfigData = @{
    AllNodes = @(
        @{
            NodeName = 'localhost'
            TargetRebootTime = (Get-Date 0:0:0.0).AddDays(3)
            PSDscAllowDomainUser = $true
            PsDscRunAsCredential = $localadmin
        }
    )
}

UpdateAndScheduleReboot -ConfigurationData $ConfigData -OutputPath (Join-Path -Path @($env:SystemDrive) -ChildPath ('Temp\UpdateAndScheduleReboot'))
Start-DscConfiguration -Path C:\temp\UpdateAndScheduleReboot -Force -Wait -Verbose

#install-module -Name @('xComputerManagement','xPendingReboot','xWindowsUpdate') -Force
```

**Comment by Chirishman** [comment]
TL;DR scripts in their default context run as `NT AUTHORITY\SYSTEM` _don't_ run elevated according to UAC which breaks crap.

**Comment by johlju** [comment]
Can you see if the resource [ScheduledTask](https://github.com/PowerShell/ComputerManagementDsc/wiki/ScheduledTask) can help you?

**Comment by Chirishman** [comment]
The ScheduledTask resource is similar but distinct from PS Scheduled jobs and this isn't the only UAC issue that I've run into when trying to use the script resource.

As it is the script resource is of fairly limited utility. The community members who I've asked have suggested writing a custom DSC resource any time you would consider using the script resource.

**Comment by johlju** [comment]
> Unfortunately I am somehow getting an elevation related error when the below configuration hits the Register-ScheduledJob step in the SetScript.

Running as a user creates the schedule job successfully. Example script below.

```powershell
Configuration TestRegisterScheduledJob
{
    [CmdletBinding()]
    param
    (
        [Parameter(Mandatory = $true)]
        [System.Management.Automation.PSCredential]
        [System.Management.Automation.Credential()]
        $Credential
    )

    Import-DscResource -ModuleName PSDscResources

    [DateTime] $TargetRebootTime = '18:00:00'

    Node localhost
    {
        User 'CreateUserAccount'
        {
            Ensure   = 'Present'
            UserName = Split-Path -Path $Credential.UserName -Leaf
            Password = $Credential
        }

        Group 'AddUserToAdministrators'
        {
            GroupName        = 'Administrators'
            Ensure           = 'Present'
            MembersToInclude = @($Credential.UserName)
        }

        Script RebootEnableJob
        {
            GetScript  = {
                return @{Result = 'HelloWorld'}
            }

            TestScript = {
                $ScheduledEnableJob = Get-ScheduledJob | Where-Object -FilterScript {$_.Name -eq 'EnableReboot'}
                if ($ScheduledEnableJob)
                {
                    return $true
                }
                else
                {
                    return $false
                }
            }

            SetScript  = {
                $EnableJob = @{
                    Name               = 'EnableReboot'
                    ScriptBlock        = {Start-Sleep -Second 30}
                    Trigger            = New-JobTrigger -Once -At $Using:TargetRebootTime
                    ScheduledJobOption = New-ScheduledJobOption -RunElevated -DoNotAllowDemandStart
                    ErrorAction        = 'Stop'
                }

                Register-ScheduledJob @EnableJob
            }

            PsDscRunAsCredential = $Credential
        }
    }
}
```

Result:
```
PS > Get-ScheduledJob

Id         Name            JobTriggers     Command                                  Enabled
--         ----            -----------     -------                                  -------
1          EnableReboot    1               Start-Sleep -Second 30                   True
```
